File: actions_cpp/src/move_robot_client.cpp
#include "rclcpp/rclcpp.hpp"
#include "rclcpp_action/rclcpp_action.hpp"
#include "my_robot_interfaces/action/move_robot.hpp"
#include "example_interfaces/msg/bool.hpp"

using MoveRobot = my_robot_interfaces::action::MoveRobot;
using MoveRobotGoalHandle = rclcpp_action::ClientGoalHandle<MoveRobot>;
using namespace std::placeholders;

class MoveRobotClientNode : public rclcpp::Node {
public:
    MoveRobotClientNode() : Node("move_robot_client") {
        move_robot_client_ = rclcpp_action::create_client<MoveRobot>(this, "move_robot");
        move_robot_subscriber_ = this->create_subscription<example_interfaces::msg::Bool>("cancel_move", 10,
            std::bind(&MoveRobotClientNode::cancel_callback, this, _1));
    }

    void send_goal(int position, int velocity) {
        move_robot_client_->wait_for_action_server();

        auto goal = MoveRobot::Goal();
        goal.position = position;
        goal.velocity = velocity;

        auto options = rclcpp_action::Client<MoveRobot>::SendGoalOptions();
        options.goal_response_callback = std::bind(&MoveRobotClientNode::goal_response_callback, this, _1);
        options.feedback_callback = std::bind(&MoveRobotClientNode::feedback_callback, this, _1, _2);
        options.result_callback = std::bind(&MoveRobotClientNode::goal_result_callback, this, _1);

        RCLCPP_INFO(this->get_logger(), "Sending a goal...");
        move_robot_client_->async_send_goal(goal, options);
    }

private:
    void cancel_callback(const example_interfaces::msg::Bool::SharedPtr msg) {
        if (msg->data && goal_handle_) {
            RCLCPP_INFO(this->get_logger(), "Cancel the goal");
            move_robot_client_->async_cancel_goal(goal_handle_);
            goal_handle_.reset();
        }
    }   
    void goal_response_callback(const MoveRobotGoalHandle::SharedPtr &goal_handle) {
        if (!goal_handle) {
            RCLCPP_WARN(this->get_logger(), "Goal got rejected");
        }
        else {
            this-> goal_handle_ = goal_handle;
            RCLCPP_INFO(this->get_logger(), "Goal got accepted");
        }
    }

    void feedback_callback(const MoveRobotGoalHandle::SharedPtr &goal_handle,
        const std::shared_ptr<const MoveRobot::Feedback> feedback) {
        (void)goal_handle;
        int position = feedback->current_position;
        RCLCPP_INFO(this->get_logger(), "Robot is at: %d", position);
    }

    void goal_result_callback(const MoveRobotGoalHandle::WrappedResult &result) {
        auto status = result.code;
        if (status == rclcpp_action::ResultCode::SUCCEEDED) {
            RCLCPP_INFO(this->get_logger(), "Succeeded");
        }
        else if (status == rclcpp_action::ResultCode::ABORTED) {
            RCLCPP_ERROR(this->get_logger(), "Aborted");
        }
        else if (status == rclcpp_action::ResultCode::CANCELED) {
            RCLCPP_WARN(this->get_logger(), "Cancelled");
        }

        int robot_position = result.result->position;
        RCLCPP_INFO(this->get_logger(), "Final Position: %d", robot_position);
    }

    rclcpp_action::Client<MoveRobot>::SharedPtr move_robot_client_;
    rclcpp::Subscription<example_interfaces::msg::Bool>::SharedPtr move_robot_subscriber_;
    MoveRobotGoalHandle::SharedPtr goal_handle_;
};


int main(int argc, char **argv) {
    rclcpp::init(argc, argv);
    auto node = std::make_shared<MoveRobotClientNode>();
    node->send_goal(77, 7);
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}

File: actions_py/actions_py/move_robot_client.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from rclpy.action import ActionClient
from rclpy.action.client import ClientGoalHandle, GoalStatus
from my_robot_interfaces.action import MoveRobot
from my_robot_interfaces.msg import CancelMove


class MoveRobotClientNode(Node): # MODIFY NAME
    def __init__(self):
        super().__init__("move_robot_client") # MODIFY NAME
        self.goal_handle_ = None
        self.subscriber_ = self.create_subscription(CancelMove, "cancel_move", self.cancel_goal, 10)
        self.move_robot_client_ = ActionClient(self, MoveRobot, "move_robot")
    
    def send_goal(self, position, velocity):
        # Wait for server
        # self.move_robot_client_.wait_for_server()
        if not self.move_robot_client_.wait_for_server(timeout_sec=5.0):
            self.get_logger().error("Action server not available!")
            return
        
        # Create a goal
        goal = MoveRobot.Goal()
        goal.position = position
        goal.velocity = velocity
        
        # Send the goal
        self.get_logger().info("Sending the goal")
        self.move_robot_client_. \
            send_goal_async(goal, feedback_callback=self.goal_feedback_callback). \
            add_done_callback(self.goal_response_callback)
        
    def cancel_goal(self, msg):
        if self.goal_handle_ is not None and self.goal_handle_.status == GoalStatus.STATUS_EXECUTING:
            self.get_logger().info("Sending a cancel request")
            self.goal_handle_.cancel_goal_async()
    
    def goal_response_callback(self, future):
        self.goal_handle_: ClientGoalHandle = future.result()
        if self.goal_handle_.accepted:
            self.get_logger().info("Goal got accepted")
            self.goal_handle_.get_result_async().add_done_callback(self.goal_result_callback)
        else:
            self.get_logger().warn("Goal got rejected")
    
    def goal_result_callback(self, future):
        status = future.result().status
        result = future.result().result
        if status == GoalStatus.STATUS_SUCCEEDED:
            self.get_logger().info("Success")
        elif status == GoalStatus.STATUS_ABORTED:
            self.get_logger().error("Aborted")
        elif status == GoalStatus.STATUS_CANCELED:
            self.get_logger().warn("Canceled")
        
        self.get_logger().info(f"Position: {result.position}")
        self.get_logger().info(f"Message: {result.message}")
    
    def goal_feedback_callback(self, feedback_msg):
        current_position = feedback_msg.feedback.current_position
        self.get_logger().info(f"Got feedback: {current_position}")

def main(args=None):
    rclpy.init(args=args)
    node = MoveRobotClientNode() # MODIFY NAME
    node.send_goal(76, 7)
    rclpy.spin(node)
    rclpy.shutdown()


if __name__ == "__main__":
    main()
